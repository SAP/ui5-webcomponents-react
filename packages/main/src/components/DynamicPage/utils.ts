export const DynamicPageCssVariables = {
  headerDisplay: '--_ui5wcr_DynamicPage_header_display',
  titleFontSize: '--_ui5wcr_DynamicPage_title_fontsize'
};
