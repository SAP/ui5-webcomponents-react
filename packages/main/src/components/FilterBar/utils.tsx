import { cloneElement } from 'react';

const inputTagNames = new Set([
  'UI5-COMBOBOX',
  'UI5-DATE-PICKER',
  'UI5-DATERANGE-PICKER',
  'UI5-DATETIME-PICKER',
  'UI5-DURATION-PICKER',
  'UI5-INPUT',
  'UI5-MULTI-INPUT',
  'UI5-RATING-INDICATOR',
  'UI5-STEP-INPUT',
  'UI5-TIME-PICKER'
]);

export const filterValue = (ref, child) => {
  const tagName = ref.tagName;
  let filterItemProps = {};
  if (inputTagNames.has(tagName)) {
    filterItemProps = { value: ref.value };
  }
  if (tagName === 'UI5-SELECT' || tagName === 'UI5-MULTI-COMBOBOX') {
    const selectedIndices = Array.from(ref.children as HTMLCollectionOf<any>)
      .map((item, index) => (item.selected ? index : false))
      .filter((el) => el !== false);
    const selectedIndicesSet = new Set(selectedIndices);
    const options = child.props.children.props.children?.map((item, index) => {
      if (selectedIndicesSet.has(index)) {
        return cloneElement(item, { selected: true });
      }
      return cloneElement(item, { selected: false });
    });
    filterItemProps = { children: options };
  }
  if (tagName === 'UI5-SWITCH' || tagName === 'UI5-CHECKBOX') {
    filterItemProps = { checked: ref.checked };
  }
  return filterItemProps;
};

export const renderSearchWithValue = (renderSearchElement, searchValue) => {
  return cloneElement(renderSearchElement, { value: searchValue ?? renderSearchElement?.props?.value });
};
