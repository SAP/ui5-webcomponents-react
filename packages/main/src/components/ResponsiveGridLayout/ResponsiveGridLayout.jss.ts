export const ResponsiveGridLayoutStyles = {
  container: {
    display: 'grid'
  },
  [`@media (max-width: 599px)`]: {
    container: {
      gridTemplateColumns: 'repeat(var(--_ui5wcr-rgl-columns-s), minmax(0, 1fr))',
      '& > *': {
        gridColumn: 'span var(--_ui5wcr-rgl-column-span-s)'
      }
    }
  },
  [`@media (min-width: 600px) and (max-width: 1023px)`]: {
    container: {
      gridTemplateColumns: 'repeat(var(--_ui5wcr-rgl-columns-m), minmax(0, 1fr))',
      '& > *': {
        gridColumn: 'span var(--_ui5wcr-rgl-column-span-m)'
      }
    }
  },
  [`@media (min-width: 1024px) and (max-width: 1439px)`]: {
    container: {
      gridTemplateColumns: 'repeat(var(--_ui5wcr-rgl-columns-l), minmax(0, 1fr))',
      '& > *': {
        gridColumn: 'span var(--_ui5wcr-rgl-column-span-l)'
      }
    }
  },
  [`@media (min-width: 1440px)`]: {
    container: {
      gridTemplateColumns: 'repeat(var(--_ui5wcr-rgl-columns-xl), minmax(0, 1fr))',
      '& > *': {
        gridColumn: 'span var(--_ui5wcr-rgl-column-span-xl)'
      }
    }
  }
};
