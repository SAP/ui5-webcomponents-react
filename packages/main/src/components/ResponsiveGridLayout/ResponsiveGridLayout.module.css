.container {
  display: grid;
}

@media (max-width: 599px) {
  .container {
    grid-template-columns: repeat(var(--_ui5wcr-rgl-columns-s), minmax(0, 1fr));

    > * {
      grid-column: span var(--_ui5wcr-rgl-column-span-s);
    }
  }
}

@media (min-width: 600px) and (max-width: 1023px) {
  .container {
    grid-template-columns: repeat(var(--_ui5wcr-rgl-columns-m), minmax(0, 1fr));

    > * {
      grid-column: span var(--_ui5wcr-rgl-column-span-m);
    }
  }
}

@media (min-width: 1024px) and (max-width: 1439px) {
  .container {
    grid-template-columns: repeat(var(--_ui5wcr-rgl-columns-l), minmax(0, 1fr));

    > * {
      grid-column: span var(--_ui5wcr-rgl-column-span-l);
    }
  }
}

@media (min-width: 1440px) {
  .container {
    grid-template-columns: repeat(var(--_ui5wcr-rgl-columns-xl), minmax(0, 1fr));

    > * {
      grid-column: span var(--_ui5wcr-rgl-column-span-xl);
    }
  }
}
