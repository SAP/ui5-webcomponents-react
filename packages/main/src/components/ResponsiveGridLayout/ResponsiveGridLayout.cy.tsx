import { ResponsiveGridLayout, ResponsiveGridLayoutPropTypes } from './index';
import { cypressPassThroughTestsFactory } from '@/cypress/support/utils';

const TestComp = (props: ResponsiveGridLayoutPropTypes) => (
  <ResponsiveGridLayout data-testid="rgl" {...props}>
    <div style={{ height: '100px', background: 'lightblue' }} data-testid={'0'} />
    <div style={{ height: '100px', background: 'cadetblue' }} data-testid={'1'} />
    <div style={{ height: '100px', background: 'skyblue' }} data-testid={'2'} />
    <div style={{ height: '100px', background: 'lightskyblue' }} data-testid={'3'} />
    <div style={{ height: '100px', background: 'blueviolet' }} data-testid={'4'} />
    <div style={{ height: '100px', background: 'cornflowerblue' }} data-testid={'5'} />
    <div style={{ height: '100px', background: 'aliceblue' }} data-testid={'6'} />
    <div style={{ height: '100px', background: 'darkblue' }} data-testid={'7'} />
    <div style={{ height: '100px', background: 'deepskyblue' }} data-testid={'8'} />
    <div style={{ height: '100px', background: 'lightsteelblue' }} data-testid={'9'} />
    <div style={{ height: '100px', background: 'mediumblue' }} data-testid={'10'} />
    <div style={{ height: '100px', background: 'mediumslateblue' }} data-testid={'11'} />
    <div style={{ height: '100px', background: 'blue' }} data-testid={'12'} />
    <div style={{ height: '100px', background: 'midnightblue' }} data-testid={'13'} />
    <div style={{ height: '100px', background: 'powderblue' }} data-testid={'14'} />
    <div style={{ height: '100px', background: 'royalblue' }} data-testid={'15'} />
    <div style={{ height: '100px', background: 'steelblue' }} data-testid={'16'} />
    <div style={{ height: '100px', background: 'slateblue' }} data-testid={'17'} />
    <div style={{ height: '100px', background: 'dodgerblue' }} data-testid={'18'} />
  </ResponsiveGridLayout>
);
describe('ResponsiveGridLayout', () => {
  it('span & columns', () => {
    cy.viewport(1440, 1080);
    cy.mount(<TestComp />);
    cy.findByTestId('rgl').should(
      'have.css',
      'grid-template-columns',
      `81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px 81.5px`
    );

    cy.viewport(1439, 1080);
    cy.mount(<TestComp />);
    cy.findByTestId('rgl').should(
      'have.css',
      'grid-template-columns',
      `111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px 111.25px`
    );
    cy.viewport(1023, 1080);
    cy.mount(<TestComp />);
    cy.findByTestId('rgl').should(
      'have.css',
      'grid-template-columns',
      `118.875px 118.875px 118.875px 118.875px 118.875px 118.875px 118.875px 118.875px`
    );
    cy.viewport(599, 1080);
    cy.mount(<TestComp />);
    cy.findByTestId('rgl').should('have.css', 'grid-template-columns', `139.75px 139.75px 139.75px 139.75px`);

    cy.mount(
      <TestComp
        columnsXL={10}
        columnsL={10}
        columnsM={10}
        columnsS={10}
        columnSpanXL={5}
        columnSpanL={5}
        columnSpanM={5}
        columnSpanS={5}
      />
    );

    cy.findByTestId('rgl').then(($el) => {
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-columns-xl')).equal('10');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-columns-l')).equal('10');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-columns-m')).equal('10');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-columns-s')).equal('10');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-column-span-s')).equal('5');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-column-span-m')).equal('5');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-column-span-l')).equal('5');
      expect(getComputedStyle($el.get(0)).getPropertyValue('--_ui5wcr-rgl-column-span-xl')).equal('5');
    });
  });

  it('gaps', () => {
    cy.mount(<TestComp />);
    cy.findByTestId('rgl').should('have.css', 'grid-row-gap', '8px').should('have.css', 'grid-column-gap', '8px');
    cy.mount(<TestComp rowGap="1rem" columnGap="1rem" />);
    cy.findByTestId('rgl').should('have.css', 'grid-row-gap', '16px').should('have.css', 'grid-column-gap', '16px');
  });

  it('custom span', () => {
    cy.mount(
      <ResponsiveGridLayout data-testid="rgl">
        <div style={{ height: '100px', background: 'lightblue', gridColumn: 'span 2' }} data-testid={'0'} />
        <div style={{ height: '100px', background: 'cadetblue', gridColumn: 'span 5' }} data-testid={'1'} />
        <div style={{ height: '100px', background: 'skyblue' }} data-testid={'2'} />
      </ResponsiveGridLayout>
    );
    cy.findByTestId('0').should('have.css', 'grid-column', 'span 2 / auto');
    cy.findByTestId('1').should('have.css', 'grid-column', 'span 5 / auto');
    cy.findByTestId('2').should('have.css', 'grid-column', 'span 1 / auto');
  });

  cypressPassThroughTestsFactory(ResponsiveGridLayout);
});
