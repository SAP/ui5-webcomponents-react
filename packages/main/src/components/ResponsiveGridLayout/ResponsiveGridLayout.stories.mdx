import { Canvas, Meta, Story } from '@storybook/addon-docs';
import { ArgsTableWithNote } from '@docs/ArgsTableWithNote';
import { ResponsiveGridLayout } from '@ui5/webcomponents-react';
import { DocsHeader } from '@docs/DocsHeader';
import { Footer } from '@docs/Footer';
import { DefaultStory } from './ResponsiveGridLayout.stories';

<Meta
  title="Layouts & Floorplans / ResponsiveGridLayout"
  component={ResponsiveGridLayout}
  argTypes={{ children: { control: { disable: true } } }}
/>

<DocsHeader />

<br />

## Example

<Canvas>
  <Story name="Default" story={DefaultStory}></Story>
</Canvas>

## Properties

<ArgsTableWithNote story="." />

<Footer />
