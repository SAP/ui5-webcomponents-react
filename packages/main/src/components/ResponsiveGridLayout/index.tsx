import { clsx } from 'clsx';
import React, { CSSProperties, forwardRef, ReactNode } from 'react';
import { createUseStyles } from 'react-jss';
import { CommonProps } from '../../interfaces/CommonProps';
import { ResponsiveGridLayoutStyles } from './ResponsiveGridLayout.jss';

export interface ResponsiveGridLayoutPropTypes extends CommonProps {
  /**
   * Number of columns to show on small screens (`max-width: 599px`)
   */
  columnsS?: number;

  /**
   * Number of columns to show on medium screens (`width >= 600px and width <=1023px`)
   */
  columnsM?: number;

  /**
   * Number of columns to show on large screens (`width >= 1024px and width <=1439px`)
   */
  columnsL?: number;

  /**
   * Number of columns to show on extra large screens (`min-width: 1440px`)
   */
  columnsXL?: number;

  /**
   * Defines how many columns a single child should cover on small screens.
   */
  columnSpanS?: number;
  /**
   * Defines how many columns a single child should cover on medium screens.
   */
  columnSpanM?: number;
  /**
   * Defines how many columns a single child should cover on large screens.
   */
  columnSpanL?: number;
  /**
   * Defines how many columns a single child should cover on extra large.
   */
  columnSpanXL?: number;
  /**
   * Gap between two columns of the grid
   */
  columnGap?: '0.5rem' | '1rem';

  /**
   * Gap between two rows of the grid
   */
  rowGap?: '0.5rem' | '1rem';

  /**
   * Child elements to be placed inside of the grid. In case you want to control the column span of each child independently,
   * you can add an element style using the [gridColumn shorthand](https://developer.mozilla.org/en-US/docs/Web/CSS/grid-column).
   *
   * Example: If you want one child to span across 4 columns, you can add this element style to the child element:
   * `style={{ gridColumn: 'span 4' }}`
   */
  children?: ReactNode | ReactNode[];
}

const useStyles = createUseStyles(ResponsiveGridLayoutStyles, { name: 'ResponsiveGridLayout' });

/**
 * The Responsive Grid Layout positions UI elements in a multi column flow layout.
 * They can be configured to display a variable number of columns depending on available screen size.
 * With these controls, it is possible to achieve flexible layouts and line breaks for large, medium,
 * and small-sized screens, such as desktop, tablet, and mobile.
 *
 * @since 0.16.4
 */
const ResponsiveGridLayout = forwardRef<HTMLDivElement, ResponsiveGridLayoutPropTypes>((props, ref) => {
  const {
    children,
    columnGap,
    rowGap,
    columnsS,
    columnsM,
    columnsL,
    columnsXL,
    columnSpanS,
    columnSpanM,
    columnSpanL,
    columnSpanXL,
    style,
    className,
    ...rest
  } = props;
  const classes = useStyles();
  const finalClassNames = clsx(classes.container, className);

  return (
    <div
      ref={ref}
      className={finalClassNames}
      style={
        {
          gridRowGap: rowGap,
          gridColumnGap: columnGap,
          '--_ui5wcr-rgl-columns-s': columnsS,
          '--_ui5wcr-rgl-columns-m': columnsM,
          '--_ui5wcr-rgl-columns-l': columnsL,
          '--_ui5wcr-rgl-columns-xl': columnsXL,
          '--_ui5wcr-rgl-column-span-s': columnSpanS,
          '--_ui5wcr-rgl-column-span-m': columnSpanM,
          '--_ui5wcr-rgl-column-span-l': columnSpanL,
          '--_ui5wcr-rgl-column-span-xl': columnSpanXL,
          ...style
        } as CSSProperties
      }
      {...rest}
    >
      {children}
    </div>
  );
});

ResponsiveGridLayout.displayName = 'ResponsiveGridLayout';

ResponsiveGridLayout.defaultProps = {
  columnGap: '0.5rem',
  rowGap: '0.5rem',
  columnsS: 4,
  columnsM: 8,
  columnsL: 12,
  columnsXL: 16,
  columnSpanS: 1,
  columnSpanM: 1,
  columnSpanL: 1,
  columnSpanXL: 1
};

export { ResponsiveGridLayout };
