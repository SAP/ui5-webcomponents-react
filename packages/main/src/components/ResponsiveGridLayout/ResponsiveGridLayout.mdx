import { ControlsWithNote, DocsHeader, Footer } from '@sb/components';
import { Canvas, Meta } from '@storybook/blocks';
import * as ComponentStories from './ResponsiveGridLayout.stories';

<Meta of={ComponentStories} />

<DocsHeader since="0.16.4" />

<br />

## Example

<Canvas of={ComponentStories.Default} />

## Properties

<ControlsWithNote of={ComponentStories.Default} />

<Footer />
