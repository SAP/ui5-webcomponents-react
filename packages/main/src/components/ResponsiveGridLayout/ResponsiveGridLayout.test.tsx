import { screen, render } from '@shared/tests';
import { createPassThroughPropsTest } from '@shared/tests/utils';
import { ResponsiveGridLayout } from '@ui5/webcomponents-react/dist/ResponsiveGridLayout';
import React from 'react';

describe('ResponsiveGridLayout', () => {
  test('Render without Crashing', () => {
    const { asFragment } = render(
      <ResponsiveGridLayout>
        <div style={{ gridColumn: 'span 2' }} data-testid={'span2'} />
        <div style={{ gridColumn: 'span 3' }} data-testid={'span3'} />
        <div style={{ gridColumn: 'span 4' }} />
        <div style={{ gridColumn: 'span 5' }} />
        <div data-testid={'defaultspan'} />
        <div />
        <div />
        <div style={{ gridColumn: 'span 4' }} />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
        <div />
      </ResponsiveGridLayout>
    );

    expect(screen.getByTestId('span2')).toHaveStyle(`grid-column: span 2`);
    expect(screen.getByTestId('span3')).toHaveStyle(`grid-column: span 3`);
    // can't assert on default-span because jsdom does not support media queries at the moment: https://github.com/jsdom/jsdom/issues/2805
    // expect(screen.getByTestId('defaultspan')).toHaveStyle('grid-column: span 1');

    expect(asFragment()).toMatchSnapshot();
  });

  createPassThroughPropsTest(ResponsiveGridLayout);
});
