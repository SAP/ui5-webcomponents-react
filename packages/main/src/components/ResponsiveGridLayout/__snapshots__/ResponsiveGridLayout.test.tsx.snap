// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResponsiveGridLayout Render without Crashing 1`] = `
<DocumentFragment>
  <div
    class="ResponsiveGridLayout-container"
    style="--_ui5wcr-rgl-columns-s: 4; --_ui5wcr-rgl-columns-m: 8; --_ui5wcr-rgl-columns-l: 12; --_ui5wcr-rgl-columns-xl: 16; --_ui5wcr-rgl-column-span-s: 1; --_ui5wcr-rgl-column-span-m: 1; --_ui5wcr-rgl-column-span-l: 1; --_ui5wcr-rgl-column-span-xl: 1;"
  >
    <div
      data-testid="span2"
      style="grid-column: span 2;"
    />
    <div
      data-testid="span3"
      style="grid-column: span 3;"
    />
    <div
      style="grid-column: span 4;"
    />
    <div
      style="grid-column: span 5;"
    />
    <div
      data-testid="defaultspan"
    />
    <div />
    <div />
    <div
      style="grid-column: span 4;"
    />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;
